### PR TITLE
feat: Add Nix config that installs rustup

### PR DESCRIPTION
**Pull request by smanilov**
This is an attempt to share my local shell.nix which installs rustup.

I got it from here: https://nixos.wiki/wiki/Rust#Installation_via_rustup

Nix instructions: https://coda.io/d/Metacraft-Labs-Wiki_db0u3vDsy39/Nix-Devshells_suHgW?searchClick=c0adf925-91d4-4fcf-8437-172565dbd833_b0u3vDsy39#_lu-La

### CODE REVIEW

**Comment by smanilov** [comment]
@PetarKirov : as the Nix expert, how does this look for you? It works for me on my machine, but I'm not sure how else to test this.
@nickysn : is this something we actually want, or do you prefer to keep the Nix config out of the repo? I think we should rather have it, since we use NixOS on the work machines.

File: shell.nix
Comment (smanilov) [inline shell.nix:1]: Note: this is a verbatim copy from the Nix + Rust page from the commit message.

**Comment by nickysn** [comment]
> @nickysn : is this something we actually want, or do you prefer to keep the Nix config out of the repo? I think we should rather have it, since we use NixOS on the work machines.

The only potential problem I see is that it creates an inconvenience for users, who use other distros and want to create their own `.envrc`. It's not a big deal, though.

**Comment by smanilov** [comment]
> > @nickysn : is this something we actually want, or do you prefer to keep the Nix config out of the repo? I think we should rather have it, since we use NixOS on the work machines.
> 
> The only potential problem I see is that it creates an inconvenience for users, who use other distros and want to create their own `.envrc`. It's not a big deal, though.

Aaah, thanks, good point, I didn't think of that.

**Comment by smanilov** [comment]
> @PetarKirov : as the Nix expert, how does this look for you? It works for me on my machine, but I'm not sure how else to test this. @nickysn : is this something we actually want, or do you prefer to keep the Nix config out of the repo? I think we should rather have it, since we use NixOS on the work machines.

Asked @melatron to manually test on his machine and he got this error:

```
Compiling const-oid v0.9.4                                                                                                                                                                                                                                                                       │
│error: linking with `cc` failed: exit status: 1                                                                                                                                                                                                                                                     │
│  |                                                                                                                                                                                                                                                                                                 │
│  = note: LC_ALL="C" PATH="/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/self-contained:/home/adikov/.rustup/toolchains/nightl│
│y-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/self-contained:/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/b│
│in:/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/self-contained:/nix/store/nbad47q0m0m9c5xid7zh05hiknwircbp-patchelf-0.15.0/bin:/nix/store/9bv7dcvmfcjnmg5mnqwqlq2wxfn8d7yi-gcc-wrapper-13.2.0/bin:/nix/store/14c6s4xzhy14i2b05s00rjns2j│
│93gzz4-gcc-13.2.0/bin:/nix/store/c2i631h8i5vcs1sqifwxfsazhwrg6wr5-glibc-2.39-52-bin/bin:/nix/store/php4qidg2bxzmm79vpri025bqi0fa889-coreutils-9.5/bin:/nix/store/kln7kinji3b7sz8r50h4gn9yy6k1js9a-binutils-wrapper-2.41/bin:/nix/store/bgcaxhhxswzvmxjbbgvvaximm5hwghz1-binutils-2.41/bin:/nix/store│
│/mzhqknx2mc94jdz4n320hn1lml86398y-clang-wrapper-17.0.6/bin:/nix/store/gx27b857v63sm7syli75gjzdvayf2xsr-clang-17.0.6/bin:/nix/store/7bd6mlnyj4192ssw50hksnqbxmj98p26-binutils-wrapper-2.41/bin:/nix/store/as8kvinr3k93sml6r77czxcyac98giy8-llvm-binutils-wrapper-17.0.6/bin:/nix/store/ncw864rylz6x57│
│l3x20fsf8gg1h05bfa-llvm-binutils-17.0.6/bin:/nix/store/xkf4p6bkcsvpxr49882w0vc9ckaf7zl9-rustup-1.26.0/bin:/nix/store/2ryq88lrrrrmidsvridcv8rcs3arp1f3-compiler-rt-libc-17.0.6/bin:/nix/store/php4qidg2bxzmm79vpri025bqi0fa889-coreutils-9.5/bin:/nix/store/jjcsr5gs4qanf7ln5c6wgcq4sn75a978-findutil│
│s-4.9.0/bin:/nix/store/i34mknsjgrfyy71k2h79gda0bvagzc2j-diffutils-3.10/bin:/nix/store/5zjms21vpxlkbc0qyl5pmj2sidfmzmd7-gnused-4.9/bin:/nix/store/28gpmx3z6ss3znd7fhmrzmvk3x5lnfbk-gnugrep-3.11/bin:/nix/store/8vvkbgmnin1x2jkp7wcb2zg1p0vc4ks9-gawk-5.2.2/bin:/nix/store/rik7p68cq7yzlj5pmfpf4yv6jnr│
│pvlgf-gnutar-1.35/bin:/nix/store/j5chw7v1x3vlmf3wmdpdb5gwh9hl0b80-gzip-1.13/bin:/nix/store/mxcq77rlan82dzpv3cgj0fh6qvv8ncil-bzip2-1.0.8-bin/bin:/nix/store/cdzpn0rdq810aknww3w9fy3wmw9ixr66-gnumake-4.4.1/bin:/nix/store/306znyj77fv49kwnkpxmb0j2znqpa8bj-bash-5.2p26/bin:/nix/store/0lfxbmchigx9vs9│
│qmrlbahcy6nxwfnj1-patch-2.7.6/bin:/nix/store/6i4xxaa812vsbli9jkq4mksdddrk27lw-xz-5.4.6-bin/bin:/nix/store/xx7x1dwybpssfhq8yikvzz38bh3yrq97-file-5.45/bin:/home/adikov/code/repos/noir/.direnv/bin:/run/wrappers/bin:/home/adikov/.nix-profile/bin:/nix/profile/bin:/home/adikov/.local/state/nix/pro│
│file/bin:/etc/profiles/per-user/adikov/bin:/nix/var/nix/profiles/default/bin:/run/current-system/sw/bin:/home/adikov/.cargo/bin:/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/" VSLANG="1033" "cc" "-m64" "/tmp/rustcemOirq/symbols.o" "/home/adikov/code/repos/noir/target/d│
│ebug/build/tiny-keccak-8e97957583556b97/build_script_build-8e97957583556b97.build_script_build.c0c35adfd172ecc7-cgu.0.rcgu.o" "/home/adikov/code/repos/noir/target/debug/build/tiny-keccak-8e97957583556b97/build_script_build-8e97957583556b97.cm9ixpsia6dsz5qcrzkgfzgu8.rcgu.o" "-Wl,--as-needed" │
│"-L" "/home/adikov/code/repos/noir/target/debug/deps" "-L" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/lib│
│std-6701759d8343e913.rlib" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-9ff13545d1688b3c.rlib" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-21│
│0d920812faea91.rlib" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-f3d3451767410a17.rlib" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-1a79dd36d08│
│251de.rlib" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-886230e7120831b2.rlib" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-2caea079085a58a2│
│.rlib" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-314a4689716cb02e.rlib" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-5727477b0a78105a.rlib│
│" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-7e555563aa211118.rlib" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-d9ed94365│
│2059c0d.rlib" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-ee5b5774583426df.rlib" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-c3e1f40644ffc9b8.rlib"│
│ "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-e8bfe52be756260a.rlib" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-7b90a2705bcf265b.rlib" "/home/adikov│
│/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-b6892f3c52c68f01.rlib" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-f72b956e24d1de70.rlib" "/home/a│
│dikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-632ae0f28c5e55ff.rlib" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-e8b7e96e438f08f6.rlib" "-Wl,-Bdyna│
│mic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-B/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld" "-B/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-│
│ld" "-B/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld" "-fuse-ld=lld" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "│
│-L" "/home/adikov/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/self-contained" "-o" "/home/adikov/code/repos/noir/target/debug/build/tiny-keccak-8e97957583556b97/build_script_build-8e97957583556b97" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,n│
│ow" "-nodefaultlibs"                                                                                                                                                                                                                                                                                │
│  = note: collect2: fatal error: execvp: No such file or directory                                                                                                                                                                                                                                  │
│          compilation terminated. 
```

I think what is needed is some sort of `binutils` package also added to the `flake.nix`, but I will think about it a little more carefully...

**Comment by smanilov** [comment]
Not quite related to this PR, but the repo cannot be built on @spartakvarna's Mac Pro M1. It fails with the following bizarre error:

```
Compiling plotters-svg v0.3.5
error[E0282]: type annotations needed for Box<_>
  --> /Users/dilyandokov/.cargo/registry/src/index.crates.io-6f17d22bba15001f/time-0.3.25/src/formatdescription/parse/mod.rs:83:9
   |
83 |     let items = format_items
   |         ^^^^^
...
86 |     Ok(items.into())
   |              ---- type must be known at this point
   |
help: consider giving items an explicit type, where the placeholders ` are specified
   |
83 |     let items: Box<_> = format_items
   |              ++++++++

   Compiling debugid v0.8.0
For more information about this error, try rustc --explain E0282.
error: could not compile time (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
warning: noirc_evaluator` (lib) generated 2 warnings
```

We tried to reproduce by building time 0.3.25 from source with the same toolchain, but we could not.

**Comment by smanilov** [comment]
This should work now on the desktop Nix work machines.

**Pull request merged**